### PR TITLE
Don't block the message loop on Docker builds

### LINKED ISSUE #313: Docker builds on Python appear to hang
The way we do Docker builds in Python blocks the message loop and leads to a pretty poor experience. Especially for large images, the initial stack doesn't even show up in the CLI, never mind the actual Image object itself. It also means incremental build output simply doesn't show up. This experience gives the impression of a hang.

We should be building asynchronously rather than blocking.

### PR DESCRIPTION
**Pull request by joeduffy**
This swaps out subprocess.Popen and communicate calls, which
block, when invoking the Docker CLI, and instead uses
asyncio.create_subprocess_exec, which offers equivalent functionlity
without blocking the message loop.

This addresses the responsiveness issues and behaves better with
the Pulumi CLI. In particular, for lengthy builds, we no longer
interfere with the ability of the CLI to report progress and display
resources as they are created by the program, asynchronously
interleaved amongst the build itself.

This is technically a breaking change, however, because the
functions now return futures where they used to return raw values.
It would be reasonable to, instead of changing these in place,
turn them into _async routines (which are used internally), and
then have thin wrappers that block alongside them.

This change fixes #313.

### CODE REVIEW

File: sdk/python/pulumi_docker/docker.py
Comment (lukehoban) [inline sdk/python/pulumi_docker/docker.py:660]: `cmd` is now no longer used, right?

**Review by lukehoban — APPROVED** [review]
LGTM. 

I was surprised at first that there weren’t any changes needed in image.py, but it looks like the code there already behaves as needed when the updated functions return futures. I’m not worried about the breaking changes in the helper functions, those are intended to be internal implementation details, with the `Image` class the only public API (even though this quite possibly isn’t enforced as strictly as it could be).

Aside - if we didn’t want to make an even bigger change (#132), it would be nice to use MLC support to avoid having 4 implementations of this with subtly different concurrency behaviour.

**Comment by lukehoban** [comment]
BTW - I’m sort of glad to see the failing test here - I was asking @stack72 last week about why we hadn’t seen test failures on those tests that are validating output after the buildkit changes in Docker. It looks like they are finally showing up here (but that we aren’t running any nightly/weekly tests in this repo?).  We will likely need to do a baseline update to reflect what Docker is doing now, pending deeper improvements related to buildkit changes.

File: sdk/python/pulumi_docker/docker.py
Comment (joeduffy) [reply sdk/python/pulumi_docker/docker.py:660]: Indeed! Good catch, fixed.

**Comment by joeduffy** [comment]
> BTW - I’m sort of glad to see the failing test here - I was asking @stack72 last week about why we hadn’t seen test failures on those tests that are validating output after the buildkit changes in Docker. It looks like they are finally showing up here (but that we aren’t running any nightly/weekly tests in this repo?). We will likely need to do a baseline update to reflect what Docker is doing now, pending deeper improvements related to buildkit changes.

I have to admit, I'm a little confused about why only the .NET tests are failing. I haven't had the chance to dive deeper yet. It actually just looks like some escape codes or something. I also verified it happens without my change.

@stack72 Are we running these tests nightly or weekly? I am embarrassed to admit, I don't know how to check.

**Comment by stack72** [comment]
> > BTW - I’m sort of glad to see the failing test here - I was asking @stack72 last week about why we hadn’t seen test failures on those tests that are validating output after the buildkit changes in Docker. It looks like they are finally showing up here (but that we aren’t running any nightly/weekly tests in this repo?). We will likely need to do a baseline update to reflect what Docker is doing now, pending deeper improvements related to buildkit changes.
> 
> 
> 
> I have to admit, I'm a little confused about why only the .NET tests are failing. I haven't had the chance to dive deeper yet. It actually just looks like some escape codes or something. I also verified it happens without my change.
> 
> 
> 
> @stack72 Are we running these tests nightly or weekly? I am embarrassed to say, I don't know how to check.

@joeduffy we don't (currently) have a cron here to run the jobs but I spoke to @lukehoban about adding one - don't block on your PR here, I'll have to follow up with some test fixes but go ahead

**Comment by lukehoban** [comment]
>  don't block on your PR here, I'll have to follow up with some test fixes but go ahead

Opened https://github.com/pulumi/pulumi-docker/issues/315 to track fixing up the tests to pass in master.

**Comment by joeduffy** [comment]
Personally, I never like merging a change without a green test run. So, despite this providing a clear user benefit, I am happy to wait for the tests to be passing before we merge. Let me know if/how I can help with that!

File: examples/util.go
Comment (lukehoban) [inline examples/util.go:30]: Is this sha really stable across build environments?

**Pull request merged**
